### DISCUSSION TITLE
Usage Example

### DISCUSSION THREAD

**Discussion by Mr-Craig** [AndreaCatania/RuntimeSkeletalMeshGenerator#4] [Q&A] [answered]
Hello,

Are you able to provide an example of how you have used this class to generate the skeletal mesh? I've been trying to implement this into my game but I keep getting crashes during the PostLoad function. I think it might be the way I'm importing the mesh but I'm not too sure.

![](https://i.imgur.com/M7Fxvc7.png)

**Comment by AndreaCatania** [comment]
Hey!

```
		USkeletalMesh* SkeletalMesh= NewObject<USkeletalMesh>();
		SkeletalMesh->RefSkeleton = Skeleton->GetReferenceSkeleton();
		SkeletalMesh->Skeleton = Skeleton;
		GenerateSkeletalMesh(
			SkeletalMesh,
			Surfaces,
			SurfacesMaterial);
```

This is how we use it, and remember that both `Surfaces` and `SurfacesMaterial` should be both fully filled with proper data. From your error, it seems you have some issues at Indices level.

**Comment by Mr-Craig** [comment]
The indices seems to be correct, i think it's the verticies that are being incorrectly imported. How are you filling your surface data, if you don't mind showing me an example?

**Comment by AndreaCatania** [comment, marked as answer]
Sorry but it's a bit too complex our use case to easily share it in an helpful manner.

Few things to be aware off; each surface has all the data relative to the surface itself:
- The vertex indices are relative to the surface, meaning that each surface starts from 0. 
- Some vertices may be duplicate across many surfaces, since the surface has all the data to fully build itself.

Recently, I've added an function that perform the decomposition of a `USkeletalMesh` [FRuntimeSkeletalMeshGenerator::DecomposeSkeletalMesh](https://github.com/AndreaCatania/RuntimeSkeletalMeshGenerator/blob/main/Source/RuntimeSkeletalMeshGenerator/RuntimeSkeletalMeshGenerator.cpp#L655) check it to get some context on how the surface array is supposed to be created.

If you figure out the problem, please consider to suggest some documentation, and if you keep having issues please post a sample.

**Reply by Mr-Craig** [reply]
Thanks for getting back to me, i actually deleted the code i previously written. I re-wrote it with the plugin and it appears to work now, I'm not too sure what was the problem! Here is how I'm using this plugin:
```
USkeletalMesh* UFBXSkeletalMeshRuntime::OpenFile(USkeleton* BaseSkeleton, UMaterialInterface* material) 
{
	IDesktopPlatform* fpl = FDesktopPlatformModule::Get();

	TArray<FString> Files;
	fpl->OpenFileDialog(0, "Open 3D Model", "", "", "FBX (*.fbx)|*.fbx| DAE (*.dae)|*.dae", 0, Files);

	if (!Files.IsValidIndex(0)) 
		return nullptr;

	FString FileString = Files.Last(0);

	if (FileString.IsEmpty()) { 
		return nullptr;
	}

	Assimp::Importer importer;

	const aiScene* Scene = NULL;

	importer.SetPropertyBool(AI_CONFIG_IMPORT_FBX_PRESERVE_PIVOTS, false);
	Scene = importer.ReadFile(TCHAR_TO_ANSI(*FileString), aiProcess_Triangulate | aiProcess_CalcTangentSpace | aiProcess_MakeLeftHanded | aiProcess_FlipUVs | aiProcess_FixInfacingNormals | aiProcess_FindDegenerates | aiProcess_FindInvalidData);

	if (!Scene) {
		return nullptr;
	}

	TArray<FMeshSurface> meshSurfaces;

	for (uint32 m = 0; m < Scene->mNumMeshes; m++) {
		aiMesh* Mesh = Scene->mMeshes[m];

		FMeshSurface mSurface;

		mSurface.Vertices.SetNum(Mesh->mNumVertices);
		mSurface.Normals.SetNum(Mesh->mNumVertices);
		mSurface.Tangents.SetNum(Mesh->mNumVertices);
		mSurface.BoneInfluences.SetNum(Mesh->mNumVertices);
		mSurface.Uvs.SetNum(Mesh->mNumVertices);

		for (uint32 v = 0; v < Mesh->mNumVertices; v++) {

			mSurface.Vertices[v] = FVector(Mesh->mVertices[v].x, -Mesh->mVertices[v].y, -Mesh->mVertices[v].z);

			if (Mesh->HasNormals())
				mSurface.Normals[v] = FVector(Mesh->mNormals[v].x, -Mesh->mNormals[v].y, -Mesh->mNormals[v].z);
			else
				mSurface.Normals[v] = FVector::ZeroVector;

			if (Mesh->HasTangentsAndBitangents())
				mSurface.Tangents[v] = FVector(Mesh->mTangents[v].x, -Mesh->mTangents[v].y, -Mesh->mTangents[v].z);
			else
				mSurface.Tangents[v] = FVector::ZeroVector;

			mSurface.FlipBinormalSigns.Add(false);

			TArray<FVector2D> UVs;

			for (uint32 uv = 0; uv < AI_MAX_NUMBER_OF_TEXTURECOORDS; uv++) {
				if (Mesh->HasTextureCoords(uv))
					UVs.Add(FVector2D(Mesh->mTextureCoords[uv][v].x, Mesh->mTextureCoords[uv][v].y));
			}

			if (UVs.Num() <= 0) {
				UVs.Add(FVector2D::ZeroVector);
			}

			mSurface.Uvs[v] = UVs;

			TArray<FRawBoneInfluence> boneInfluences;
			if (Mesh->HasBones()) {
				for (uint32 b = 0; b < Mesh->mNumBones; b++) {
					aiBone* Bone = Mesh->mBones[b];

					for (uint32 w = 0; w < Bone->mNumWeights; w++) {
						aiVertexWeight Weight = Bone->mWeights[w];

						FRawBoneInfluence boneInf;

						boneInf.BoneIndex = b;
						boneInf.VertexIndex = Weight.mVertexId;
						boneInf.Weight = Weight.mWeight;

						if (boneInf.VertexIndex == v) {
							boneInfluences.Add(boneInf);
						}
					}
				}
			}
			mSurface.BoneInfluences[v] = boneInfluences;
		}

		for (uint32 f = 0; f < Mesh->mNumFaces; f++) {
			aiFace Face = Mesh->mFaces[f];

			mSurface.Indices.Add(Face.mIndices[0]);
			mSurface.Indices.Add(Face.mIndices[1]);
			mSurface.Indices.Add(Face.mIndices[2]);
		}

		meshSurfaces.Add(mSurface);
	}

	USkeletalMesh* skeletalMesh = NewObject<USkeletalMesh>();
	skeletalMesh->RefSkeleton = BaseSkeleton->GetReferenceSkeleton();
	skeletalMesh->Skeleton = BaseSkeleton;

	TArray<UMaterialInterface*> Materials;

	for (auto surface : meshSurfaces)
	{
		Materials.Add(material);
	}

	FRuntimeSkeletalMeshGenerator::GenerateSkeletalMesh(
		skeletalMesh,
		meshSurfaces,
		Materials
	);

	return skeletalMesh;
}
```

Thanks again for the help and this great plugin!

**Reply by AndreaCatania** [reply]
Welcome, I'm glad it works now, most likely there was a bug in the previous lib version. However, if I can give you an hint, instead of adding the surface mesh to the array, you can initialize the array with the correct size and compose the array directly, in this way you can avoid that the `FMeshSurface` is copied.
```c++
TArray<FMeshSurface> meshSurfaces(Scene->mNumMeshes);
for (uint32 m = 0; m < Scene->mNumMeshes; m++) {
    //...
    FMeshSurface& mSurface = meshSurfaces[m];
    //...
```
Something like this ---^
Of course, you don't need any longer `meshSurfaces.Add(mSurface);`. 🍻

**Reply by baymaxzq111** [reply]
USkeletalMesh* UFBXSkeletalMeshRuntime::OpenFile(USkeleton* BaseSkeleton, UMaterialInterface* material)，   USkeleton* BaseSkeletonIs this the imported integral mesh?

**Comment by Art3m1s-GITHUB** [comment]
Dear Developer I have a question for you. I tried the method proposed by @Mr-Craig witch I used as a blueprint function. The method proposed returns a SkeletalMesh. Now, how can I replace the character mesh with the mesh just loaded from file during runtime?
Thank you in advance for your reply.
Best regards.

**Reply by Mr-Craig** [reply]
This is how i had to setup when i was testing it, i just checked if the return was valid and called set skeletal mesh on the skeletal mesh component

![UE4Editor_iG4n6ny1mz](https://user-images.githubusercontent.com/9359117/152632052-2b389ebf-455c-4bad-9024-883f481fb63b.jpg)

**Reply by AndreaCatania** [reply]
@Art3m1s-GITHUB Yup, `SetSkeletalMesh` is what you need, @Mr-Craig thanks a lot!!

**Reply by Art3m1s-GITHUB** [reply]
> This is how i had to setup when i was testing it, i just checked if the return was valid and called set skeletal mesh on the skeletal mesh component
> 
> ![UE4Editor_iG4n6ny1mz](https://user-images.githubusercontent.com/9359117/152632052-2b389ebf-455c-4bad-9024-883f481fb63b.jpg)

When I add the SetSkeletalMesh the application crashes and I receive this error: Assertion failed: NumElements <= NumInlineElements

**Reply by Mr-Craig** [reply]
> When I add the SetSkeletalMesh the application crashes and I receive this error: Assertion failed: NumElements <= NumInlineElements

are you able to send the full stack trace with symbols?

**Comment by Fransferdy** [comment]
Hello, thanks for providing this plugin! Though currently I'm failing to do a simple asset loading example, I wonder if I could get some help.

I tried to load a skeletal mesh to an actor component, using the code(and blueprints) provided by @Mr-Craig , using UE 4.27.2, and two things keep happening, any complex FBX (such as the UE Mannequin), UE crashes, and simpler models won't show(a cube with a single bone).
Crashes happens at these lines (With Unreadable memory exception):
`aiVertexWeight Weight = Bone->mWeights[w];`
if I comment the bone part, it crashes here
`mSurface.Normals[v] = FVector(Mesh->mNormals[v].x, -Mesh->mNormals[v].y, -Mesh->mNormals[v].z);`

Then I tried to load simpler models, I made a Cube in blender with a single bone in the middle, I passed the Mannequin base skeleton and a material from the Sample Starter in blueprints. In this case it doesn't crash, but during playtest, nothing appears where the model should be.

I thought that maybe the Assimp I compiled could be the culprit, so I made a simple importer for .OFF 3d model files, I made a function similar to the one @Mr-Craig provided and tried to load .OFF models with SkeletalMesh, but again, it does not crash, but nothing appear where the model should be displayed.

And I'm not exactly sure where things are going wrong. The same issue happens with @Mr-Craig code, and with mine. I wonder if the problem is with the 3D models I'm trying to load, or  with the plugin or something else entirely.

Here's my .OFF importer code(https://en.wikipedia.org/wiki/OFF_(file_format)):
(Basically I just stripped @Mr-Craig method, and mixed it with a .off importer I took from other tool)

```
USkeletalMesh* USkeletalMeshLoader::OpenFile(USkeleton* BaseSkeleton, UMaterialInterface* material){

	std::string filepath = "C:\\dev\\projects\\monsterGame\\voxelizer\\mesh-voxelization\\examples\\raw\\1.off";
	std::ifstream* file = new std::ifstream(filepath.c_str());
	std::string line;
	std::stringstream ss;
	int line_nb = 0;

	bool usecolor = false;

	std::getline(*file, line);
	++line_nb;
	
	if (line != "off" && line != "OFF") {
		std::cout << "[Error] Invalid header: \"" << line << "\", " << filepath << std::endl;
		return false;
	}

	size_t n_edges;
	std::getline(*file, line);
	++line_nb;

	int n_vertices;
	int n_faces;
	ss << line;
	ss >> n_vertices;
	ss >> n_faces;
	ss >> n_edges;

	UE_LOG(LogTemp, Warning, TEXT("LOADING MODEL, %d %d %d"), n_vertices, n_faces, n_edges);

	TArray<FMeshSurface> meshSurfaces;

	FMeshSurface mSurface;

	mSurface.Vertices.SetNum(n_vertices);
	mSurface.Normals.SetNum(n_vertices);
	mSurface.Tangents.SetNum(n_vertices);
	mSurface.BoneInfluences.SetNum(n_vertices);
	mSurface.Uvs.SetNum(n_vertices);

	for (size_t v = 0; v < n_vertices; ++v) {
		std::getline(*file, line);
		++line_nb;

		ss.clear();
		ss.str("");

		FVector vertex;
		float x, y, z;
		ss << line;
		ss >> x;
		ss >> y;
		ss >> z;

		mSurface.Vertices[v] = FVector(x, y, z);
		mSurface.Normals[v] = FVector::ZeroVector;
		mSurface.Tangents[v] = FVector::ZeroVector;

		mSurface.FlipBinormalSigns.Add(false);

		TArray<FVector2D> UVs;
		UVs.Add(FVector2D::ZeroVector);
		mSurface.Uvs[v] = UVs;

		TArray<FRawBoneInfluence> boneInfluences;
		FRawBoneInfluence boneInf;

		boneInf.BoneIndex = 0;
		boneInf.VertexIndex = v;
		boneInf.Weight = 1;

		boneInfluences.Add(boneInf);
		mSurface.BoneInfluences[v] = boneInfluences;
	}

	for (size_t f = 0; f < n_faces; ++f) {
		std::getline(*file, line);
		++line_nb;

		ss.clear();
		ss.str("");

		size_t n;
		ss << line;
		ss >> n;

		if (n != 3) {
			std::cout << "[Error] Not a triangle (" << n << " points) at " << (line_nb - 1) << std::endl;
			return false;
		}

		//std::cout << "Empty color vector: " << color(0) <<  " " << color(1) << " " << color(2) << std::endl;
		int32 i1, i2, i3;
		int color1, color2, color3;

		ss >> i1;
		ss >> i2;
		ss >> i3;

		if (usecolor == true) {
			ss >> color1;
			ss >> color2;
			ss >> color3;
		};
		//std::cout << "Color vector after reading line: " << color(0) <<  " " << color(1) << " " << color(2) << std::endl;

		mSurface.Indices.Add(i1);
		mSurface.Indices.Add(i2);
		mSurface.Indices.Add(i3);

		//mesh.add_color(color);
	}


	file->close();
	delete file;

	meshSurfaces.Add(mSurface);
	USkeletalMesh* skeletalMesh = NewObject<USkeletalMesh>();
	skeletalMesh->SetRefSkeleton(BaseSkeleton->GetReferenceSkeleton());
	skeletalMesh->SetSkeleton(BaseSkeleton);

	TArray<UMaterialInterface*> Materials;

	for (auto surface : meshSurfaces)
	{
		Materials.Add(material);
	}

	FRuntimeSkeletalMeshGenerator::GenerateSkeletalMesh(
		skeletalMesh,
		meshSurfaces,
		Materials
	);

	return skeletalMesh;
}
```

Any help is appreciated.
I would also be very happy if I could see a sample project where this is currently working : )

**Reply by Fransferdy** [reply]
I was able to import a Pyramid with an improved version of my Off importer, I'll be continuing from there, no more need of help, thanks.

**Reply by AndreaCatania** [reply]
Glad you found a solution, let me know if you need any help! Also you can use this as reference to know how to create a SkeletalMesh.

**Reply by Zydzzb** [reply]
@Fransferdy 
Hello, I successfully imported the SkeletalMesh using the OFF file as you instructed. How can I ensure that the materials are applied correctly? The model I imported doesn't have any materials.

**Comment by Fransferdy** [comment]
I just uploaded my full sample (the farthest I went with this) here:
https://github.com/Fransferdy/creaturecreator/tree/main

But basically, if I remember well, your skeletal mesh is formed of several "surfaces" each surface is basically a "model", in other words, vertexes, uvs, bone indexes weights. when you call GenerateSkeletalMesh you pass it three parameters,
GenerateSkeletalMesh(
		skeletalMesh,
		meshSurfaces,
		Materials
	);
skeletalMesh will be your result, after the mesh is created, meshSurfaces is basically your model, divided into "surfaces" sections, and Materials is an array of materials, that will apply to each surface. So I believe, Material[0] will apply to Surface[0], Material[1] will apply to Surface[1] and so on.

From that project
```
	meshSurfaces.Empty();
	mSurface.Vertices.Empty();
	mSurface.Normals.Empty();
	mSurface.Tangents.Empty();
	mSurface.BoneInfluences.Empty();
	mSurface.Uvs.Empty();
	mSurface.Colors.Empty();
	mSurface.FlipBinormalSigns.Empty();
```
```
USkeletalMesh* USkeletalMeshLoader::GenerateBlob(TArray<UMaterialInterface*> mats) {

	UE_LOG(LogTemp, Warning, TEXT("Meshes: %d"), meshSurfaces.Num());
	if (meshSurfaces.Num() == 0)
	{
		GenerateMyBlob();
	}


	USkeletalMesh* skeletalMesh = NewObject<USkeletalMesh>();

	USkeleton* skeleton = NewObject<USkeleton>();
	buildSkeleton((FReferenceSkeleton&)skeleton->GetReferenceSkeleton(), skeleton);

	FString name = skeleton->GetReferenceSkeleton().GetBoneName(0).ToString();
	FString name2 = skeleton->GetReferenceSkeleton().GetBoneName(1).ToString();

	UE_LOG(LogTemp, Warning, TEXT("Bone name %s , %s"), *name, *name2);

	skeletalMesh->SetRefSkeleton(skeleton->GetReferenceSkeleton());
	skeletalMesh->SetSkeleton(skeleton);


	FRuntimeSkeletalMeshGenerator::GenerateSkeletalMesh(
		skeletalMesh,
		meshSurfaces,
		mats
	);

	return skeletalMesh;
}
```

**Reply by Zydzzb** [reply]
@Fransferdy 

Thank you very much for your response; it has been very helpful. It's evident that you are a dedicated and tech-savvy individual. I have another question, and I'm wondering if you are familiar with Unreal Engine 5's Runtime IK Retargeting feature. My current requirement is as follows: to be able to import a SkeletalMesh with the same bone hierarchy as the default Unreal character at runtime and establish a bone mapping between the two. This way, when I drive the default Unreal character, I can use the bone chain to control the locally uploaded character. Of course, this is just my vision. Do you think this is achievable? After uploading, does the SkeletalMesh retain the same bone hierarchy as the default character?

Thank you very much for your assistance!

**Reply by Fransferdy** [reply]
I would like to point out that currently this plugin only works with UE 5.1/5.0, with 5.2 onwards it breaks(there seems to have been a lot of changes with UE code).
I am a complete amateur when it comes to Unreal Engine :) 
Now, I think it is possible what you just asked. For example, you could pass a Skeletal Mesh to your function:
USkeletalMesh* USkeletalMeshLoader::GenerateBlob(TArray<UMaterialInterface*> mats, SkeletalMesh *myUEMannequin)
and then "take" the skeleton from that SkeletalMesh, something like

uploadedSkeletalMesh->SetRefSkeleton(myUEMannequin->GetReferenceSkeleton());
uploadedSkeletalMesh->SetSkeleton(myUEMannequin.getSkeleton());

I'm not sure it would work, since I haven't tried that myself, but I think it would, if the skeleton you imported has the exact same bone hierarchy. You could also map, as you said, if you can get myUEMannequin.getSkeleton(), you might as well, find the bones position, and do some math, not sure how this would go. I know that the editor has a feature in which you associate similar skeletons to use the same animations within them, I used that to put paragon animations on the mannequin and metahumans, if you can do it in the editor, it is probably possible to do it with cpp code, though it is an adventure to find out how haha
